sysconfig/cloudinit.go

```diff
@@ -88,3 +92,220 @@ func configureCloudInit(opts *Options) (err error) {
+
+	cloudInitSnapdRestrictFile = "/etc/cloud/cloud.cfg.d/zzzz_snapd.cfg"
+
+	nocloudRestrictYaml = []byte(`datasource_list: [NoCloud]
+datasource:
+  NoCloud:
+    fs_label: null`)
+
+	genericCloudRestrictYamlPattern = `datasource_list: [%s]`
+)
+
+const (
+	// CloudInitDisabledPermanently is when cloud-init is disabled as per the
+	// cloud-init.disabled file.
+	CloudInitDisabledPermanently CloudInitState = iota
+	// CloudInitRestrictedBySnapd is when cloud-init has been restricted by
+	// snapd with a specific config file.
+	CloudInitRestrictedBySnapd
+	// CloudInitUntriggered is when cloud-init is disabled because nothing has
+	// triggered it to run, but it could still be run.
+	CloudInitUntriggered
+	// CloudInitDone is when cloud-init has been run on this boot.
+	CloudInitDone
+	// CloudInitEnabled is when cloud-init is active, but not necessarily
+	// finished. This matches the "running" and "not run" states from cloud-init
+	// as well as any other state that does not match any of the other defined
+	// states, as we are conservative in assuming that cloud-init is doing
+	// something.
+	CloudInitEnabled
+	// CloudInitErrored is when cloud-init tried to run, but failed or had invalid
+	// configuration.
+	CloudInitErrored
+)
+
+// CloudInitStatus returns the current status of cloud-init. Note that it will
+// first check for static file-based statuses first through the snapd
+// restriction file and the disabled file before consulting
+// cloud-init directly through the status command.
+// Also note that in unknown situations we are conservative in assuming that
+// cloud-init may be doing something and will return CloudInitEnabled when we
+// do not recognize the state returned by the cloud-init status command.
+func CloudInitStatus() (CloudInitState, error) {
+	// if cloud-init has been restricted by snapd, check that first
+	snapdRestrictingFile := filepath.Join(dirs.GlobalRootDir, cloudInitSnapdRestrictFile)
+	if osutil.FileExists(snapdRestrictingFile) {
+		return CloudInitRestrictedBySnapd, nil
+	}
+
+	// if it was explicitly disabled via the cloud-init disable file, then
+	// return special status for that
+	disabledFile := filepath.Join(dirs.GlobalRootDir, "etc/cloud/cloud-init.disabled")
+	if osutil.FileExists(disabledFile) {
+		return CloudInitDisabledPermanently, nil
+	}
+
+	out, err := exec.Command("cloud-init", "status").CombinedOutput()
+	if err != nil {
+		return CloudInitErrored, osutil.OutputErr(out, err)
+	}
+	// output should just be "status: <state>"
+	match := cloudInitStatusRe.FindSubmatch(out)
+	if len(match) != 2 {
+		return CloudInitErrored, fmt.Errorf("invalid cloud-init output: %v", osutil.OutputErr(out, err))
+	}
+	switch string(match[1]) {
+	case "disabled":
+		// here since we weren't disabled by the file, we are in "disabled but
+		// could be enabled" state - arguably this should be a different state
+		// than "disabled", see
+		// https://bugs.launchpad.net/cloud-init/+bug/1883124 and
```

Comment:
@mvo5  Good bug refs . I'll keep this snapd consumer in mind if we change cloud-init status output . I expect eventually you'd like to consume `cloud-init status --long` to get a more detailed determination of why cloud-init is currently disabled.